Always render when XR is enabled, even if no OS windows can draw

dsnopek opened the pull request:
**The short explanation:**

Currently, in `Main::iteration()`, it will skip rendering for a particular frame if `DisplayServer::get_singleton()->can_any_window_draw()` returns `false`.

For all current `DisplayServer`s, they either:

1. Always return `true`, or
2. Return `true` if at least one of Godot's windows is visible

However, when using an XR device, we need to keep rendering, even if all OS windows aren't visible. For example, if you have a VR headset connected to your desktop, you don't want it to stop rendering to the headset if the Godot window on your desktop is minimized.

So, this PR will cause rendering to happen if the primary `XRInterface` is initialized, even if `DisplayServer::get_singleton()->can_any_window_draw()` returns `false`.

**The longer explanation:**

I started looking into this because there is a case when using the Meta XR Simulator on MacOS, where we would sometimes call `xrWaitFrame()` twice without calling  `xrBeginFrame()` in between, which leads to the application getting "stuck" forever (spinning beach ball and everything).

According the OpenXR specification ([in Section 10.3](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#frame-synchronization)):

> A subsequent [xrWaitFrame](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#xrWaitFrame) call must block until the previous frame has been begun with [xrBeginFrame](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#xrBeginFrame) and must unblock independently of the corresponding call to [xrEndFrame](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#xrEndFrame).

So, the Meta XR Simulator is acting exactly as specified, and is blocking until the next `xrBeginFrame()` (which will never come when Godot is using single-threaded rendering).

The reason the `xrBeginFrame()` sometimes doesn't happen, is that we're calling it at the beginning of rendering (as we should, per the spec!), and, per the "short explanation" above, Godot may skip rendering if all windows are momentarily not visible.

While I originally dug into this for this very specific OpenXR related bug, I think it logically makes sense that when using XR, we shouldn't make rendering conditional on if OS windows are visible.

dsnopek committed:
Always render when XR is enabled, even if no OS windows can draw

clayjohn commented on main/main.cpp:
```diff
@@ -4108,8 +4108,20 @@ bool Main::iteration() {
+
+#ifndef _3D_DISABLED
+	if (!can_draw) {
+		Ref<XRInterface> xr_interface = XRServer::get_singleton()->get_primary_interface();
+		if (xr_interface.is_valid() && xr_interface->is_initialized()) {
```
Is it possible for this condition to change at run time, or can this be cached?

My preference would be for the value to be cached in the DisplayServer and then the XRInterface  would essentially be treated as any other window and affect the result of `can_any_window_draw()`

dsnopek replied:
> Is it possible for this condition to change at run time, or can this be cached?

This can change from frame-to-frame, if the application enters or exits XR mode.

> My preference would be for the value to be cached in the DisplayServer and then the XRInterface would essentially be treated as any other window and affect the result of `can_any_window_draw()`

Hm. The DisplayServer is platform dependent, ie we have `DisplayServerWindows`, `DisplayServerX11`, `DisplayServerMacOS`, etc, whereas this check is platform independent.

I suppose we could put a function on the base class that combines the `XRServer` check with the platform specific check? But it still feels better to me to have the `XRServer` check outside of `DisplayServer`, to keep the platform dependent/independent split very clean.

BastiaanOlij replied:
Great find David but I agree with Clay that this get very convoluted especially if we have other processes that want to overrule this. I can think of other use cases where we may want to render as per normal even if the application is minimized (think of our movie output, I might start that to generate a movie that last 15 minutes and minimize Godot while that is chugging away). 

I think we can turn this upside down and introduce a method on our OS class where OpenXR can register that there is another output that requires drawing. This needs to be made in such a way that multiple systems can register this simultaniously so it can't be a simple bool.

The code above would change to something like:
```
bool can_draw = DisplayServer::get_singleton()->can_any_window_draw() && OS::get_singleton()->can_any_output_draw();
```

And in OS we'd do something like:
```
bool OS::can_any_output_draw() {
  return !registered_output_draw_list.is_empty();
}

bool OS::register_output_draw(Ref<RefCounted> p_output) {
  if (!registered_output_draw_list.has(p_output)) {
    registered_output_draw_list.push_back(p_output);
  }
}

bool OS::unregister_output_draw(Ref<RefCounted> p_output) {
  if (registered_output_draw_list.has(p_output)) {
    registered_output_draw_list.erase(p_output);
  }
}
```

Or something along those lines

Then in `OpenXR::initialise` we call `register_output_draw` and in `OpenXR::uninitialize` we call `unregister_output_draw`

dsnopek commented:
Alright, I've updated this PR to implement @BastiaanOlij's idea, with a few small changes:

- I put the methods on `DisplayServer` rather than `OS`
- I renamed the methods to `register_additional_output()`, `unregister_additonal_output()` and `has_additional_outputs()`
- I used `Object *` rather than `Ref<RefCounted>`: If the system registering the output doesn't clean up after itself, with `Ref<RefCounted>` we run the risk of an object being freed when `DisplayServer` is freed, at which point most of Godot won't be operable and anything the object tries to do in its destructor could cause issues.

However, I don't particularly like this approach:

- We're leaving it up to individual `XRInterface`s to register their additional output. On the one hand, this is good, because some interfaces may not need to (for example, WebXR), but now it's another thing that `XRInterface`s need to know to do.
- We need to expose this publicly, because an `XRInterface` could be implemented from GDExtension, and so it'll need to be able to call `(un)register_additional_output()`
- This isn't the prettiest API: registering an `Object *` is a little awkward, but I don't know what would be a better identifier.

Anyway, please let me know if this looks better to you guys!

/cc @clayjohn

BastiaanOlij commented:
I agree that the `Object *` thing is a bit weird, we could also keep a counter and just increase/decrease it. A little more fault sensitive but maybe cleaner. The point is however that the existing behaviour is far too strict. XR is an example where this fails, but I don't believe its the only example. Like I said, I think we'll find other use cases for this as more and more people integrate Godot with other systems and would like it to keep working at its full ability even when minimized.

Heck a long standing wish of many users for XR is that we don't even have any Godot window open, I can see similar situations when integrating Godot into other apps.

BastiaanOlij reviewed (approved):
lgtm, it's worth having the discussion if passing Object is indeed the best way or if we simplify this to increasing/decreasing a counter

akien-mga commented on servers/display_server.h:
```diff
@@ -53,6 +53,8 @@ class DisplayServer : public Object {
 	RID _get_rid_from_name(NativeMenu *p_nmenu, const String &p_menu_root) const;
 #endif
 
+	Vector<Object *> additional_outputs;
```
Not important but can likely be a `LocalVector`.

akien-mga replied:
Maybe also of `ObjectID` instead of `Object *`? We don't reference those pointers so it's not too risky but usually the "safe" pattern for objects is to pass ObjectIDs around.

akien-mga reviewed (approved):
Looks good to me.

akien-mga commented:
Thanks!

pochoco24 commented:
I tried to capture Godot with OBS while minimized with this commit, but it still freezes in OBS when the window is minimized.

All I have in the scene is an animated moving cube, a camera, and a script:

`extends Node3D

var object = Object.new()

func _ready() -> void:
	DisplayServer.register_additional_output(object)`

dsnopek commented:
@pochoco24 I responded [here](https://github.com/godotengine/godot-proposals/issues/10286#issuecomment-2261052219)
